File: Android/Libraries/fireflydevice/src/main/java/com/fireflydesign/fireflydevice/FDVMADecoder.java
package com.fireflydesign.fireflydevice;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by denis on 5/8/15.
 */
public class FDVMADecoder implements FDPullTask.Decoder {

    public Object decode(int type, byte[] data, byte[] responseData) {
        FDBinary binary = new FDBinary(data);
        double time = binary.getUInt32(); // 4-byte time
        int interval = binary.getUInt16();
        int n = binary.getRemainingLength() / 2; // 2 bytes == sizeof(float16)
        FDFireflyDeviceLogger.info(null, "sync VMAs: %d values", n);
        List<Double> vmas = new ArrayList<Double>();
        for (int i = 0; i < n; ++i) {
            double value = binary.getFloat16();
            vmas.add(value);
        }
        FDVMAItem item = new FDVMAItem();
        item.time = time;
        item.interval = interval;
        item.vmas = vmas;
        return item;
    }

}

File: FireflyDevice/FireflyDevice/FDVMADecoder.m
//
//  FDVMADecoder.m
//  FireflyDevice
//
//  Created by Denis Bohm on 5/8/15.
//  Copyright (c) 2015 Firefly Design. All rights reserved.
//

#import "FDVMADecoder.h"

#import "FDPullTask.h"

@implementation FDVMAItem
@end

@implementation FDVMADecoder

- (id)decode:(uint32_t)type data:(NSData *)data responseData:(NSData *)responseData
{
    const int floatBytes = 2; // FDV2 uses float16
    FDBinary *binary = [[FDBinary alloc] initWithData:data];
    NSTimeInterval time = [binary getUInt32]; // 4-byte time
    uint16_t interval = [binary getUInt16];
    NSUInteger n = [binary getRemainingLength] / floatBytes; // 4 bytes == sizeof(float32), 2 bytes == sizeof(float16)
    FDFireflyDeviceLogInfo(@"FD010715", @"sync VMAs: %lu values", (unsigned long)n);
    NSMutableArray *vmas = [NSMutableArray array];
    for (NSUInteger i = 0; i < n; ++i) {
        float value = (floatBytes == 2) ? [binary getFloat16] : [binary getFloat32];
        [vmas addObject:[NSNumber numberWithFloat:value]];
    }
    FDVMAItem *item = [[FDVMAItem alloc] init];
    item.time = time;
    item.interval = interval;
    item.vmas = vmas;
    return item;
}

@end
